File: androidApp/src/main/kotlin/it/danielebufarini/trenify/presentation/RailwayFormatting.kt
package it.danielebufarini.trenify.presentation

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import it.danielebufarini.trenify.R
import it.danielebufarini.trenify.core.model.TrainCategory
import it.danielebufarini.trenify.core.model.TrainStatus
import it.danielebufarini.trenify.design.component.TrenifyStatusTone
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale

private val ROME: ZoneId = ZoneId.of("Europe/Rome")
private val TIME_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")

/**
 * Day/month formatters follow the device locale (month abbreviations are
 * user-visible words), mirroring iOS `.autoupdatingCurrent` behavior.
 * Built per call so a runtime locale change is honored; interpretation
 * stays Europe/Rome regardless of the host timezone.
 */
private fun dateFormat(): DateTimeFormatter = DateTimeFormatter.ofPattern("d MMM", Locale.getDefault())

private fun dateTimeFormat(): DateTimeFormatter = DateTimeFormatter.ofPattern("d MMM HH:mm", Locale.getDefault())

/**
 * Cross-feature railway presentation formatting. Railway wall times always
 * render in Europe/Rome, independent of the host timezone; durations stay in
 * whole minutes from shared state. No business logic lives here.
 */
fun railwayTime(epochSeconds: Long): String =
    Instant.ofEpochSecond(epochSeconds).atZone(ROME).format(TIME_FORMAT)

fun railwayDate(epochSeconds: Long): String =
    Instant.ofEpochSecond(epochSeconds).atZone(ROME).format(dateFormat())

fun railwayDateTime(epochSeconds: Long): String =
    Instant.ofEpochSecond(epochSeconds).atZone(ROME).format(dateTimeFormat())

/** Localized service-date label from an ISO-8601 date; locale medium format. */
fun railwayServiceDateLabel(iso: String): String =
    LocalDate.parse(iso).format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM))

/** Localized train-category label from shared category identity. */
@Composable
fun railwayCategoryLabel(category: TrainCategory): String = when (category) {
    TrainCategory.REG -> stringResource(R.string.st_categoryREG)
    TrainCategory.IC -> stringResource(R.string.st_categoryIC)
    TrainCategory.EC -> stringResource(R.string.st_categoryEC)
    TrainCategory.EN -> stringResource(R.string.st_categoryEN)
    TrainCategory.FR -> stringResource(R.string.st_categoryFR)
    TrainCategory.FA -> stringResource(R.string.st_categoryFA)
    TrainCategory.FB -> stringResource(R.string.st_categoryFB)
    else -> category.code
}

fun railwayDurationLabel(durationMinutes: Long): String {
    val hours = durationMinutes / 60
    val minutes = durationMinutes % 60
    return if (hours > 0) "${hours}h ${minutes.toString().padStart(2, '0')}m" else "${minutes}m"
}

/**
 * Semantic realtime tone for a railway service. Scheduled Journey legs without
 * correlated enrichment stay Unknown; a null delay never implies on-time.
 */
fun railwayStatusTone(status: TrainStatus, delayMinutes: Int?): TrenifyStatusTone = when (status) {
    TrainStatus.CANCELLED, TrainStatus.PARTIALLY_CANCELLED -> TrenifyStatusTone.Cancelled
    TrainStatus.ARRIVED -> TrenifyStatusTone.Arrived
    TrainStatus.DIVERTED, TrainStatus.RESCHEDULED -> TrenifyStatusTone.Warning
    TrainStatus.RUNNING, TrainStatus.NOT_DEPARTED -> when {
        (delayMinutes ?: 0) > 0 -> TrenifyStatusTone.Delayed
        delayMinutes == 0 -> TrenifyStatusTone.OnTime
        else -> TrenifyStatusTone.Unknown
    }
    TrainStatus.UNKNOWN -> TrenifyStatusTone.Unknown
}

File: iosApp/iosApp/Presentation/RailwayFormatting.swift
import Foundation
@preconcurrency import SharedApp

/// Cross-feature railway presentation formatting. Railway wall times always render
/// in Europe/Rome, independent of the device timezone; durations stay in
/// whole minutes from shared state. No business logic lives here.
///
/// T8.14-C2: the hot row formatters are cached shared instances instead of
/// per-call constructions (DateFormatter creation is expensive and these run
/// per row on every list render). Configuration is unchanged on purpose:
/// fixed POSIX month abbreviations stay exactly as before (T8.12 owns
/// locale decisions). All call sites run on the main thread (SwiftUI bodies
/// and main-thread XCTest); the explicit opt-out documents that confinement.
enum RailwayFormatting {
    static func date(forEpochSeconds value: Int64) -> Date {
        Date(timeIntervalSince1970: TimeInterval(value))
    }

    nonisolated(unsafe) private static let romeTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.timeZone = RomeDateConversion.romeTimeZone
        formatter.dateFormat = "HH:mm"
        return formatter
    }()

    nonisolated(unsafe) private static let romeDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.timeZone = RomeDateConversion.romeTimeZone
        formatter.dateFormat = "d MMM"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter
    }()

    nonisolated(unsafe) private static let romeDateTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.timeZone = RomeDateConversion.romeTimeZone
        formatter.dateFormat = "d MMM HH:mm"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter
    }()

    static func time(_ epochSeconds: Int64) -> String {
        romeTimeFormatter.string(from: date(forEpochSeconds: epochSeconds))
    }

    static func date(_ epochSeconds: Int64) -> String {
        romeDateFormatter.string(from: date(forEpochSeconds: epochSeconds))
    }

    static func dateTime(_ epochSeconds: Int64) -> String {
        romeDateTimeFormatter.string(from: date(forEpochSeconds: epochSeconds))
    }

    /// Localized service-date label from an ISO-8601 date; locale medium format.
    static func serviceDate(_ iso: String) -> String {
        let parts = iso.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return iso }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = RomeDateConversion.romeTimeZone
        guard let date = calendar.date(from: DateComponents(year: parts[0], month: parts[1], day: parts[2], hour: 12)) else { return iso }
        let formatter = DateFormatter()
        formatter.timeZone = RomeDateConversion.romeTimeZone
        formatter.locale = .autoupdatingCurrent
        formatter.dateStyle = .medium
        return formatter.string(from: date)
    }

    static func duration(minutes: Int64) -> String {
        let hours = minutes / 60
        let rest = minutes % 60
        if hours > 0 { return "\(hours)h \(String(format: "%02d", rest))m" }
        return "\(rest)m"
    }

    /// Semantic realtime tone for a railway service. Scheduled legs without
    /// correlated enrichment stay unknown; a null delay never implies on-time.
    static func statusTone(status: ModelTrainStatus, delayMinutes: KotlinInt?) -> TrenifyStatusTone {
        switch status {
        case .cancelled, .partiallyCancelled: return .cancelled
        case .arrived: return .arrived
        case .diverted, .rescheduled: return .warning
        case .running, .notDeparted:
            let delay = delayMinutes?.intValue ?? -1
            if delay > 0 { return .delayed }
            if delay == 0 { return .onTime }
            return .unknown
        case .unknown: return .unknown
        @unknown default: return .unknown
        }
    }
}
